File: android/app/src/main/java/com/frenchnclc7/app/ui/Themes.kt
package com.frenchnclc7.app.ui

import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.graphics.Color

/** The same colour tokens as the web app's themes (frontend/src/shared/themes.js). */
data class AppColors(
    val label: String,
    val bg: Color, val card: Color, val border: Color,
    val accent: Color, val accentSoft: Color, val onAccent: Color,
    val hard: Color, val gold: Color, val frBlue: Color, val frRed: Color,
    val text: Color, val muted: Color, val success: Color, val link: Color,
    val danger: Color, val warnText: Color, val warnSoft: Color,
)

private fun hex(v: Long) = Color(0xFF000000 or v)
private fun rgba(r: Int, g: Int, b: Int, a: Float) = Color(r / 255f, g / 255f, b / 255f, a)

object Themes {
    val all: Map<String, AppColors> = linkedMapOf(
        "midnight" to AppColors(
            "Midnight", hex(0x0B1220), hex(0x131C2E), hex(0x25314A),
            hex(0x3B82F6), rgba(59, 130, 246, 0.14f), hex(0x0B1220),
            hex(0xF59E0B), hex(0xF5B841), hex(0x2E4A9E), hex(0xB23A48),
            hex(0xE8EDF6), hex(0x8291AB), hex(0x22C55E), hex(0x93C5FD),
            hex(0xF87171), hex(0xF5C77E), rgba(245, 158, 11, 0.14f),
        ),
        "paper" to AppColors(
            "Paper", hex(0xF6F7F9), hex(0xFFFFFF), hex(0xE2E6EE),
            hex(0x2563EB), rgba(37, 99, 235, 0.10f), hex(0xFFFFFF),
            hex(0xD97706), hex(0x7F5600), hex(0x2E4A9E), hex(0xB23A48),
            hex(0x1B2233), hex(0x5B667A), hex(0x15803D), hex(0x1D4ED8),
            hex(0xB91C1C), hex(0x92400E), rgba(217, 119, 6, 0.12f),
        ),
        "lavender" to AppColors(
            "Lavender", hex(0xF3EFFB), hex(0xFFFFFF), hex(0xDDD3F0),
            hex(0x7255D6), rgba(114, 85, 214, 0.11f), hex(0xFFFFFF),
            hex(0xD97706), hex(0x7F5600), hex(0x4B3FA6), hex(0xB23A48),
            hex(0x2A1F4D), hex(0x665B8A), hex(0x15803D), hex(0x5A3FC0),
            hex(0xB91C1C), hex(0x92400E), rgba(217, 119, 6, 0.12f),
        ),
        "blush" to AppColors(
            "Blush", hex(0xFDF1F5), hex(0xFFFFFF), hex(0xF3D3DD),
            hex(0xC43570), rgba(196, 53, 112, 0.10f), hex(0xFFFFFF),
            hex(0xD97706), hex(0x7F5600), hex(0x2E4A9E), hex(0xB23A48),
            hex(0x4A1F2E), hex(0x86566A), hex(0x15803D), hex(0xA82860),
            hex(0xB91C1C), hex(0x92400E), rgba(217, 119, 6, 0.12f),
        ),
    )

    const val DEFAULT = "midnight"
    fun get(id: String?): AppColors = all[id] ?: all.getValue(DEFAULT)
}

val LocalColors = compositionLocalOf { Themes.get(Themes.DEFAULT) }

File: ios/FrenchNCLC7/Theme/Themes.swift
import SwiftUI

/// The same colour tokens as the web app's themes (frontend/src/shared/themes.js).
struct AppColors {
    let label: String
    let bg: Color, card: Color, border: Color
    let accent: Color, accentSoft: Color, onAccent: Color
    let hard: Color, gold: Color
    let text: Color, muted: Color, success: Color, link: Color, danger: Color
}

private func hex(_ value: UInt32) -> Color {
    Color(
        red: Double((value >> 16) & 0xFF) / 255,
        green: Double((value >> 8) & 0xFF) / 255,
        blue: Double(value & 0xFF) / 255
    )
}

private func rgba(_ r: Double, _ g: Double, _ b: Double, _ a: Double) -> Color {
    Color(red: r / 255, green: g / 255, blue: b / 255, opacity: a)
}

enum Themes {
    static let defaultId = "midnight"

    /// In display order.
    static let order = ["midnight", "paper", "lavender", "blush"]

    static let all: [String: AppColors] = [
        "midnight": AppColors(
            label: "Midnight", bg: hex(0x0B1220), card: hex(0x131C2E), border: hex(0x25314A),
            accent: hex(0x3B82F6), accentSoft: rgba(59, 130, 246, 0.14), onAccent: hex(0x0B1220),
            hard: hex(0xF59E0B), gold: hex(0xF5B841),
            text: hex(0xE8EDF6), muted: hex(0x8291AB), success: hex(0x22C55E), link: hex(0x93C5FD), danger: hex(0xF87171)
        ),
        "paper": AppColors(
            label: "Paper", bg: hex(0xF6F7F9), card: hex(0xFFFFFF), border: hex(0xE2E6EE),
            accent: hex(0x2563EB), accentSoft: rgba(37, 99, 235, 0.10), onAccent: hex(0xFFFFFF),
            hard: hex(0xD97706), gold: hex(0x7F5600),
            text: hex(0x1B2233), muted: hex(0x5B667A), success: hex(0x15803D), link: hex(0x1D4ED8), danger: hex(0xB91C1C)
        ),
        "lavender": AppColors(
            label: "Lavender", bg: hex(0xF3EFFB), card: hex(0xFFFFFF), border: hex(0xDDD3F0),
            accent: hex(0x7255D6), accentSoft: rgba(114, 85, 214, 0.11), onAccent: hex(0xFFFFFF),
            hard: hex(0xD97706), gold: hex(0x7F5600),
            text: hex(0x2A1F4D), muted: hex(0x665B8A), success: hex(0x15803D), link: hex(0x5A3FC0), danger: hex(0xB91C1C)
        ),
        "blush": AppColors(
            label: "Blush", bg: hex(0xFDF1F5), card: hex(0xFFFFFF), border: hex(0xF3D3DD),
            accent: hex(0xC43570), accentSoft: rgba(196, 53, 112, 0.10), onAccent: hex(0xFFFFFF),
            hard: hex(0xD97706), gold: hex(0x7F5600),
            text: hex(0x4A1F2E), muted: hex(0x86566A), success: hex(0x15803D), link: hex(0xA82860), danger: hex(0xB91C1C)
        ),
    ]

    static func get(_ id: String?) -> AppColors {
        all[id ?? defaultId] ?? all[defaultId]!
    }
}
